test/io/output/output_test.py

```diff
@@ -13,108 +13,48 @@
 
-
-class TestOpenhtf(unittest.TestCase):
+class TestOutput(unittest.TestCase):
 
   @classmethod
   def setUpClass(cls):
-    tempdir = tempfile.mkdtemp()
-    atexit.register(shutil.rmtree, tempdir)
-    conf.LoadFromDict({'station_id': 'unittest_openhtf'})
-    test = openhtf.Test(
-        dimensions, test_name='TestTest', test_description='Unittest test',
-        test_version='1.0.0')
-    test.AddOutputCallback(_CleanVariability(json_factory.OutputToJSON(
-        os.path.join(tempdir, 'record.json'), sort_keys=True, indent=2)))
-    test.AddOutputCallback(_CleanVariability(
-        mfg_inspector.OutputToTestRunProto(
-            os.path.join(tempdir, 'record.testrun'))))
-    test.Execute()
-    cls.tempdir = tempdir
-
-  def _CompareOutput(self, expected, actual):
-    # Remove any problems from trailing newlines.
-    expected, actual = expected.strip(), actual.strip()
-
-    if expected == actual:
-      return
-
-    # Output the diff first.
-    logging.error('***** TestRun mismatch:*****')
-    for line in difflib.unified_diff(
-        expected.splitlines(), actual.splitlines(),
-        fromfile='expected', tofile='actual', lineterm=''):
-      logging.error(line)
-    logging.error('^^^^^  TestRun diff  ^^^^^')
-
-    # Then raise the AssertionError as expected.
-    assert expected == actual
+    # Load input testrun from pickled file.
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                           'record.pickle'), 'rb') as picklefile:
+      cls.record = pickle.load(picklefile)
+
+    # Load our canonical 'correct' outputs from files.
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                           'record.json'), 'r') as jsonfile:
+      cls.json = jsonfile.read()
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                           'record.testrun'), 'rb') as testrunfile:
+      cls.testrun = testrun_pb2.TestRun.FromString(testrunfile.read())
 
   def testJson(self):
-    with open(os.path.join(self.tempdir, 'record.json'), 'r') as jsonfile:
-      self._CompareOutput(EXPECTED_JSON, jsonfile.read())
+    json_output = StringIO()
+    json_factory.OutputToJSON(json_output, sort_keys=True, indent=2)(self.record)
+    data.AssertEqualsAndDiff(self.json, json_output.getvalue())
 
   def testTestrun(self):
-    expected = testrun_pb2.TestRun.FromString(EXPECTED_TESTRUN)
-    with open(os.path.join(self.tempdir, 'record.testrun'), 'rb') as trfile:
-      actual = testrun_pb2.TestRun.FromString(trfile.read())
-      self._CompareOutput(
-          text_format.MessageToString(expected),
-          text_format.MessageToString(actual))
+    testrun_output = StringIO()
+    mfg_inspector.OutputToTestRunProto(testrun_output)(self.record)
+    actual = testrun_pb2.TestRun.FromString(testrun_output.getvalue())
+    data.AssertEqualsAndDiff(
+        text_format.MessageToString(self.testrun),
```

Comment:
Yeah the pickled version is pre-generated input, and isn't tested against anything.  I agree that makes it super opaque to work with outside the test.  It was generated by running the `all_the_things.py` example with a pickling output callback.  I'll update `all_the_things.py` to output that pickled record by default and add a comment here as to how to re-generate it.  I may also add a switch to apply here to run the test in "generate correct output" mode.

That still doesn't really help with making it more human readable.  The `AssertEqualsAndDiff`was intended to help a bit with that :( i think to some extent it's just the nature of testing encodings like this.  I'll do what I can to make it a little less painful.